src/transactions/builders/common.hpp

```diff
@@ -0,0 +1,200 @@
+/**
+ * This file is part of Ark Cpp Crypto.
+ *
+ * (c) Ark Ecosystem <info@ark.io>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ **/
+
+#ifndef ARK_TRANSACTIONS_BUILDER_COMMON_HPP
+#define ARK_TRANSACTIONS_BUILDER_COMMON_HPP
+
+#include <algorithm>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "common/configuration.hpp"
+
+#include "identities/keys.hpp"
+
+#include "transactions/transaction.hpp"
+
+#include "interfaces/identities.hpp"
+
+#include "utils/str.hpp"
+
+namespace Ark {
+namespace Crypto {
+namespace transactions {
+namespace builder {
+
+////////////////////////////////////////////////////////////////////////////////
+// Internal Common Transaction Builder Class
+template<class T> class Common {
+  public:
+    ////////////////////////////////////////////////////////////////////////////
+    // Version
+    T &version(const uint8_t &version) {
```

Comment:
More discussion in https://github.com/ArkEcosystem/cpp-crypto/pull/198#discussion_r357258363

Reply:
> For small types, especially if they will fit in a register (<= 64 bit), you should pass then by value instead of const &

I would pass-by-value myself because the code is shorter and in practice I think it does not matter (for small types). But why do you say "should"?

```
void f1(uint8_t x) {
    write(1, &x, sizeof(x));
}

void f2(const uint8_t& x) {
    write(1, &x, sizeof(x));
}

(gdb) disas f1
Dump of assembler code for function _Z2f1h:
   0x00000000002012e0 <+0>:	push   %rbp
   0x00000000002012e1 <+1>:	mov    %rsp,%rbp
   0x00000000002012e4 <+4>:	sub    $0x10,%rsp
   0x00000000002012e8 <+8>:	mov    %dil,-0x1(%rbp)
   0x00000000002012ec <+12>:	lea    -0x1(%rbp),%rsi
   0x00000000002012f0 <+16>:	mov    $0x1,%edx
   0x00000000002012f5 <+21>:	mov    $0x1,%edi
   0x00000000002012fa <+26>:	callq  0x2013e0 <write@plt>
   0x00000000002012ff <+31>:	add    $0x10,%rsp
   0x0000000000201303 <+35>:	pop    %rbp
   0x0000000000201304 <+36>:	retq   
End of assembler dump.
(gdb) disas f2
Dump of assembler code for function _Z2f2RKh:
   0x0000000000201310 <+0>:	push   %rbp
   0x0000000000201311 <+1>:	mov    %rsp,%rbp
   0x0000000000201314 <+4>:	mov    %rdi,%rsi
   0x0000000000201317 <+7>:	mov    $0x1,%edx
   0x000000000020131c <+12>:	mov    $0x1,%edi
   0x0000000000201321 <+17>:	pop    %rbp
   0x0000000000201322 <+18>:	jmpq   0x2013e0 <write@plt>
End of assembler dump.
```